src/RefractDataStructure.cc

```diff
@@ -756,45 +931,47 @@ namespace drafter {
     }
 
     template<typename T>
-    refract::IElement* RefractElementFromMSON(const snowcrash::DataStructure& ds)
+    refract::IElement* RefractElementFromMSON(const SectionInfo<snowcrash::DataStructure>& ds)
     {
         using namespace refract;
         typedef T ElementType;
 
         ElementType* element = new ElementType;
-        SetElementType(element, ds.typeDefinition);
+        SetElementType(element, ds.section.typeDefinition);
 
-        if (!ds.name.symbol.literal.empty()) {
-            element->meta[SerializeKey::Id] = IElement::Create(ds.name.symbol.literal);
+        if (!ds.section.name.symbol.literal.empty()) {
+            snowcrash::SourceMap<mson::Literal> sourceMap = SectionInfo<mson::Literal>::NullSourceMap();
+
+            sourceMap.sourceMap.append(ds.sourceMap.name.sourceMap);
+            
+            element->meta[SerializeKey::Id] = PrimitiveToRefract(MakeSectionInfo(ds.section.name.symbol.literal, sourceMap, ds.hasSourceMap()));
         }
 
         ElementData<T> data;
 
-        std::for_each(ds.sections.begin(), ds.sections.end(), ExtractTypeSection<T>(data, ds));
+        SectionInfoCollection<mson::TypeSections> typeSections(ds.section.sections, ds.sourceMap.sections);
 
-        TransformElementData<T>(element, data);
+        std::for_each(typeSections.begin(), typeSections.end(), ExtractTypeSection<T>(data, ds));
```

Comment:
As I said before, could have used macro here. Please check other places where the same macro could be used since I am going to stop pointing this part out.